infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java

```diff
@@ -112,9 +135,43 @@ public KubernetesEnvironment setType(String type) {
     return (KubernetesEnvironment) super.setType(type);
   }
 
-  /** Returns pods that should be created when environment starts. */
-  public Map<String, Pod> getPods() {
-    return pods;
+  /**
+   * Returns pods that should be created when environment starts.
+   *
+   * <p>Note: This map <b>should not</b> be changed, as it will only return pods and not
+   * deployments. If objects in the map need to be changed, see {@link #getPodData()}
+   */
+  public Map<String, Pod> getPodsCopy() {
+    return ImmutableMap.copyOf(pods);
+  }
+
+  /**
+   * Returns deployments that should be created when environment starts.
+   *
+   * <p>Note: This map <b>should not</b> be changed. If objects in the map need to be changed, see
+   * {@link #getPodData()}
+   */
+  public Map<String, Deployment> getDeploymentsCopy() {
+    return ImmutableMap.copyOf(deployments);
+  }
+
+  /**
+   * Returns {@link PodData} representing the metadata and pod spec of objects (pods or deployments)
+   * that should be created when environment starts. The data returned by this method represents all
+   * deployment and pod objects that form the workspace, and should be used when provisioning or
+   * performing any action that needs to see every object in the environment.
+   */
+  public Map<String, PodData> getPodData() {
+    return ImmutableMap.copyOf(podData);
+  }
+
+  /**
+   * Add a pod to the current environment. This method is necessary as the map returned by {@link
+   * #getPodsCopy()} is a copy. This method also adds the relevant data to {@link #getPodData()}.
+   */
+  public void addPod(String key, Pod pod) {
```

Comment:
I agree -- I didn't use `podName` as the parameter name mostly because I wasn't sure if that's always what was used, but it seems to be the case. I'll fix it.

Reply:
What about 
```
   public void addPod(Pod pod) {
     String podName = pod.getMetadata().getName;
     pods.put(podName, pod);
     podData.put(podName, new PodData(pod.getSpec(), pod.getMetadata()));
   }
```
?

Reply:
Good suggestion, that's a bit cleaner. Fixed.